builder/caddy-builder.sh

```diff
@@ -0,0 +1,37 @@
+#!/bin/sh
+plugins=$@
+
+mkdir -p /src/custom-caddy/cmd/caddy
+cd /src/custom-caddy/cmd/caddy
+
+cat >> main.go <<EOF
+package main
+
+import (
+	caddycmd "github.com/caddyserver/caddy/v2/cmd"
+	_ "github.com/caddyserver/caddy/v2/modules/standard"
+EOF
+
+for p in $plugins; do
+    echo -e "\t_ \"$p\"" >> main.go
+done
+
+cat >> main.go <<EOF
+)
+
+func main() {
+	caddycmd.Main()
+}
+EOF
+
+cat >> go.mod <<EOF
+module caddy
+
+go 1.14
+
+require (
+	github.com/caddyserver/caddy/v2 $CADDY_SOURCE_VERSION
+)
+EOF
```

Comment:
Any particular reason this is hard-coded (mostly)? Why not just plop down the main.go file and then run `go mod init` and then `go get ...` to pin the desired Caddy version? That will also update go.sum and download the dependencies if missing so Go can verify the build's integrity.

Reply:
@mholt it was pulling down an older version (`v2.0.0-beta9`, because `beta9` is treated as higher than `beta.15`) when I just did a `go mod init`, and `go get`ting on top of that is a bit redundant so this just saves a step.

Come to think of it, one of the reasons this image even exists is to _avoid_ pulling the caddy source down (since it's already there in `/src/caddy`), so maybe the better approach is to do a `replace github.com/caddyserver/caddy/v2 => /src/caddy` and that'll save the spurious download and pin the image to the source version.

Reply:
I am not sure if using `replace` nullifies the benefits of module security, though. We have to be careful there... we need to find out if it still verifies the checksums with `replace` being used.

Reply:
@mholt I just pushed an update - it calls `go get` now for the Caddy source, and also for each specified plugin, and I've verified that the `go.sum` is populated before the custom binary is built.

I'm no expert on how module security works, but I think we're OK on that front:
- when the Caddy source is cloned from git we specify a tag
- then the code is compiled as-is, pulling required modules from the module mirror - this will verify the checksums of Caddy dependencies
- the resulting `caddy/caddy:builder` image is now built and tagged with the git-tag of the code we asked for, and the dependencies specified in that tag's `go.sum` will have been included (and stored in the package cache in the image)
- when a user uses `caddy/caddy:builder` to build a new binary, the `caddy-builder` script will re-use the same version of the source and all the dependencies pulled down at `docker build` time
- any new modules a user adds in will then be pulled down with `go get`, which will pull it through the module mirror as well

Does that sound reasonable?

Reply:
@hairyhenderson In doing some research and talking to smart people, it seems that using `replace` should be fine. The key is that we have to trust whatever we are replacing to (RHS), but the replacement module will still be verified. So as long as you can trust whatever downloaded your replacement and added the `replace` line, it should still be safe.